### PR TITLE
WIP: Dev/risteys pipeline

### PR DESCRIPTION
**Pull request by essiviippola**
Goal: implement the Risteys pipeline for FinRegistry data

Implemented: 
- new project structure
- data loading for FinRegistry (note: wide format is used for first events instead of the long format)
- data preprocessing for FinRegistry 
- case-cohort sampling for FinRegistry (note: no FinRegistry data needed for tests)

Next steps: 
- update README.md
- prep the data for Cox proportional hazards model 
- survival & mortality analysis 
- integrate Feiyi's summary statistics
- integrate main.py and main.wdl as needed

### CODE REVIEW

File: pipeline/main.py [hunk 1]
```diff
@@ -0,0 +1,37 @@
+"""Run analyses for FinRegistry data"""
```
Comment (vincent-octo) [inline pipeline/main.py:1]: I am a bit confused here about this file fits with the config and analysis scripts and probably need some explaining :sweat_smile: .

- It appears this is FinRegistry specific?
- It is running the survival analysis only? I am guessing it is because you will focus on that first. However, would that code be better off in the survival analysis file rather than in main.py?
- The config.py doesn't seem to be used?

File: pipeline/risteys_pipeline/finregistry/load_data.py [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+)
+
+
+def load_minimal_phenotype_data(
+    data_path=FINREGISTRY_MINIMAL_PHENOTYPE_DATA_PATH, preprocess=False
+):
+    """Loads minimal phenotype data as a dataframe and optionally performs preprocessing."""
+    logger.info("Loading minimal phenotype data")
+    cols = ["FINREGISTRYID", "date_of_birth", "death_date", "sex"]
+    df = pd.read_csv(data_path, usecols=cols, header=0, sep=",")
```
Comment (vincent-octo) [inline pipeline/risteys_pipeline/finregistry/load_data.py:19]: Are `header` and `sep` needed here?

File: pipeline/risteys_pipeline/finregistry/preprocess_data.py [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+
+        Returns a dataframe with the following columns: 
+        finregistryid, date_of_birth, death_date, sex, death_age, dead, female
+    """
+    logger.info("Preprocessing minimal phenotype data")
+    df.columns = df.columns.str.lower()
+    df = df.drop_duplicates().reset_index(drop=True)
+    excluded_subjects = list_excluded_subjects(df)
+    df = df[~df["finregistryid"].isin(excluded_subjects)]
+    df["death_age"] = (df["death_date"] - df["date_of_birth"]).dt.days / 365.25
```
Comment (vincent-octo) [inline pipeline/risteys_pipeline/finregistry/preprocess_data.py:90]: Sometimes I would use variable instead of numbers, to make things more explicit.

For example here:
```python
YEAR = 365.25
...
    df["death_age"] = (df["death_date"] - df["date_of_birth"]).dt.days / YEAR
```

example: https://github.com/dsgelab/risteys/blob/78ff4af31b80c2f40f26522d3a5ca1dcf2db5544/pipeline/medication_stats_logit.py#L55-L64

File: pipeline/risteys_pipeline/finregistry/preprocess_data.py [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+    df["dead"] = ~df["death_date"].isna()
+    df["female"] = df["sex"] == 2
```
Comment (vincent-octo) [inline pipeline/risteys_pipeline/finregistry/preprocess_data.py:92]: Same here, might be better to be explicit:

```python
SEX_FEMALE = 2
...
    df["female"] = df["sex"] == SEX_FEMALE
```

File: pipeline/risteys_pipeline/finregistry/sample.py [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+    """
+    Create a sample of cases and controls and calculate case-cohort weights.
+    
+    Returns:
+        - sample_of_cases (list): finregistryids for cases
+        - sample_of_controls (list): finregistryids for controls 
+        - weight_cases (float): case-cohort weight for cases
+        - weight_controls (float): case-cohort weight for controls
+    """
+    cases = df.loc[df["outcome"] == 1, "finregistryid"].tolist()
```
Comment (vincent-octo) [inline pipeline/risteys_pipeline/finregistry/sample.py:18]: Probably not worth changing, but worth mentioning:

You can also use `df.shape` to get the size of a pandas dataframe, instead of using `df.tolist()` and then `len()` on it.
For example:
```python
cases, _ = df.loc[df["outcome"] == 1, "finregistryid"].shape
```

**Review by vincent-octo — COMMENTED** [review]
Hey @essiviippola , thanks for making your work visible here! Here is my first review :slightly_smiling_face: 

I realize this is currently a draft, and maybe I have asked questions about stuff that are only part-way there or will get there in the end. If so, feel free to ignore them for now :smile: !

I have not checked if the FinnGen scripts are still working as expected yet, not sure I'll have time before the holidays :confused: . Anyway, there are already things to discuss! 

- Could you describe what is workflow of using this new code hierarchy? For example providing the commands needed to run all the analyses using the FinnGen data, and the same for FinRegistry.

- When doing work on pandas dataframe I usually go with `df.loc[ <filters> ]` instead of the `df[ <filters> ]` you use. I kinda remember it's because it's safer to use `df.loc`, but I don't remember why. Do you know about this?

I have also added review comments in the code.

Thanks for your work! :raised_hands:

File: pipeline/.gitignore [hunk 6]
```diff
@@ -1 +1,134 @@
...
+.dmypy.json
+dmypy.json
+
+# Pyre type checker
+.pyre/
+
+# Data and notebook directories
+data/*
+notebooks/*
 *.json
```
[the remark this answers is not in the document]
Comment (essiviippola) [reply pipeline/.gitignore:134]: done: 87914c6148e745e7e502fe3a129635c28d35ccec

File: pipeline/risteys_pipeline/config.py [hunk 7]
```diff
@@ -0,0 +1,20 @@
+import os
```
[the remark this answers is not in the document]
Comment (essiviippola) [reply pipeline/risteys_pipeline/config.py]: I agree, it looks much nicer
done: eccb1b4

File: pipeline/risteys_pipeline/log.py [hunk 8]
```diff
@@ -1,12 +1,14 @@
...
 
 level = getenv("LOG_LEVEL", logging.INFO)
-logger = logging.getLogger("pipeline")
-handler = logging.StreamHandler()
 formatter = logging.Formatter(
-    "%(asctime)s %(levelname)-8s %(module)-21s %(funcName)-25s: %(message)s")
+    "%(asctime)s %(levelname)-8s %(module)-21s %(funcName)-25s: %(message)s"
+)
+# handler = logging.StreamHandler()
+handler = logging.FileHandler("risteys.log")
```
[the remark this answers is not in the document]
Comment (essiviippola) [reply pipeline/risteys_pipeline/log.py]: good point, changed it back to stream handler: ab0e668

File: pipeline/setup.py [hunk 9]
```diff
@@ -0,0 +1,3 @@
+from setuptools import setup
```
[the remark this answers is not in the document]
Comment (essiviippola) [reply pipeline/setup.py:1]: I'm following this: https://gist.github.com/ericmjl/27e50331f24db3e8f957d1fe7bbbe510. The only exceptions are the main.py and that I placed tests in a separate folder to keep them organized. I'll comment on the main.py below.

There's also this: https://packaging.python.org/en/latest/tutorials/packaging-projects/. However, I think some parts of the tutorial are not relevant to the risteys pipeline, since the python package only meant to be used by us.

File: pipeline/risteys_pipeline/config.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (essiviippola) [reply pipeline/risteys_pipeline/config.py]: hm, I think maybe a third option is to split the config into `config.ini` and `config.py` like this: 

```
# In config.ini

[finngen]
MINIMAL_PHENOTYPE_DATA_PATH = ...
WIDE_FIRST_EVENTS_DATA_PATH = ...
ENDPOINTS_DATA_PATH = ...

[finregistry]
MINIMAL_PHENOTYPE_DATA_PATH = ...
WIDE_FIRST_EVENTS_DATA_PATH = ...
ENDPOINTS_DATA_PATH = ...

[shared] # if needed
SOME_SHARED_VARIABLE = ...
```

```
# In config.py

from configparser import ConfigParser
from os import getenv

conf_dataset= getenv("CONF_DATASET", "finregistry")

config = ConfigParser()
config.read("config.ini")

MINIMAL_PHENOTYPE_DATA_PATH = config.get(conf_dataset, "MINIMAL_PHENOTYPE_DATA_PATH")
WIDE_FIRST_EVENTS_DATA_PATH = config.get(conf_dataset, "WIDE_FIRST_EVENTS_DATA_PATH")
ENDOPINTS_DATA_PATH = config.get(conf_dataset, "ENDPOINTS_DATA_PATH")
SOME_SHARED_VARIABLE = config.get("shared", "SOME_SHARED_VARIABLE")
```
And then use the configs as before, e.g. 

```
from risteys_pipeline.config import MINIMAL_PHENOTYPE_DATA_PATH
```

What do you think?

File: pipeline/main.py (hunk 1 above)
Comment (essiviippola) [reply pipeline/main.py:1]: Ehh me too, to be honest :D I wrote `main.py` to solve my problem of "I don’t actually know how the whole thing should be run yet".

Following the [DS project organization](https://gist.github.com/ericmjl/27e50331f24db3e8f957d1fe7bbbe510) everything under `risteys_pipeline` is a function, so I wrote `main.py` to run the analyses (i.e. call the functions in risteys_pipeline). Following the same guide, I think the correct place for this would actually be in `scripts/`. 

My end goal was to build a pipeline where each analysis could be run as part of a pipeline where data is preprocessed only once and stored in temporary feather files so that the same preprocessed data could be used by e.g. both "cumulative incidence" and "mortality analysis". However, based on this I should probably implement something for running each analysis independently first.

So in conclusion, I'll move all this someplace else and remove the `main.py`.

File: pipeline/risteys_pipeline/finregistry/sample.py (hunk 5 above)
Comment (essiviippola) [reply pipeline/risteys_pipeline/finregistry/sample.py]: Here `cases` is a list of finregistryids that I used later for sampling and calculating the case-cohort weights. I replaced `cases` with a better variable name `caseids`: 0b3a1ff

File: pipeline/risteys_pipeline/finregistry/preprocess_data.py (hunk 3 above)
Comment (essiviippola) [reply pipeline/risteys_pipeline/finregistry/preprocess_data.py]: Done: 5384cd8

File: pipeline/risteys_pipeline/finregistry/load_data.py (hunk 2 above)
Comment (essiviippola) [reply pipeline/risteys_pipeline/finregistry/load_data.py:19]: Not needed because they're the defaults (or the default of `header` is "infer") but I prefer them :)

File: pipeline/risteys_pipeline/finregistry/preprocess_data.py (hunk 4 above)
Comment (essiviippola) [reply pipeline/risteys_pipeline/finregistry/preprocess_data.py]: Done: 4e615a5

File: pipeline/main.py (hunk 1 above)
Comment (essiviippola) [reply pipeline/main.py]: Moved the contents of the `main.py` to `scripts/finregistry_mortality_analysis.py` (for now, at least): c5a8277

File: pipeline/risteys_pipeline/config.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (vincent-octo) [reply pipeline/risteys_pipeline/config.py]: Ah, that 3rd option looks useful!

My mind is not set on a clear way forward for the config part. I think we need first to figure out how it would look like to run the pipeline for FinnGen and for FinRegistry.
After that it should be clearer.

File: pipeline/setup.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (vincent-octo) [reply pipeline/setup.py:1]: Thanks for the resources!

I am not 100% sure, but from my understanding: since we don't need to publish the package (and actually can't get it from within the sandboxed environments we use to run pipeline), then we can simply use one (ore more) `__init__.py` to make the import work for the pipeline scripts, and we don't need the `setup.py`.
https://docs.python.org/3/tutorial/modules.html#packages

Or do you see any reason to use the setup.py for our use cases?

File: pipeline/main.py (hunk 1 above)
Comment (vincent-octo) [reply pipeline/main.py]: Sounds good :slightly_smiling_face: !

File: pipeline/setup.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (essiviippola) [reply pipeline/setup.py:1]: I think either way is good. Here's a nice (brief) description of the benefits of `setup.py` that I found today: https://goodresearch.dev/setup.html#install-a-project-package. I think for this project the main benefit would be that using the functions within notebooks (for testing/experimentation) is a bit easier.
